src/connection.js

```diff
@@ -594,15 +613,25 @@ export class Connection {
   async getBalance(
     publicKey: PublicKey,
     commitment: ?Commitment,
-  ): Promise<number> {
+  ): Promise<RpcResponseAndContext | number> {
     const args = this._argsWithCommitment([publicKey.toBase58()], commitment);
     const unsafeRes = await this._rpcRequest('getBalance', args);
-    const res = GetBalanceRpcResult(unsafeRes);
+    const res = GetBalanceAndContextRpcResult(unsafeRes);
     if (res.error) {
       throw new Error(res.error.message);
     }
     assert(typeof res.result !== 'undefined');
-    return res.result;
+    assert(typeof res.result.context !== 'undefined');
+    assert(typeof res.result.value !== 'undefined');
+
+    const rpcResponseAndContext = {
+      context: {
+        slot: res.result.context.slot,
+      },
+      value: res.result.value,
+    };
+
+    return commitment ? rpcResponseAndContext : rpcResponseAndContext.value;
```

Comment:
This conditional isn't right. Someone using this API could be relying on the default behaviour when no commitment is passed but still want to receive the context. I think we need a check above to see if the returned response has a context or not and use that to determine the return value

Reply:
There will be a few weeks until your RPC api changes are live on the next Solana release. In the meantime, we will want to use web3 sdk for examples and handle responses from both edge testnet (with your api changes) and the stable testnet (without your api changes)

Reply:
I'm having trouble seeing how that works. In this code, the client "opts in" to the new response object by providing the commitment as input. I don't see how this revision can be a drop-in replacement if we do what you're saying?

Reply:
@sunnygleason I don't see this "opt in" that you're referring to. RPC api returns the context whether you specify a commitment or not: https://github.com/solana-labs/solana/blob/master/core/src/rpc.rs#L107

Reply:
I was thinking the backward compatibility aspect would look more like what was implemented for recentBlockhash:
https://github.com/solana-labs/solana-web3.js/blob/31ecd913a4a85c20bc23a0ad8c531891870ac8c3/src/connection.js#L872
We need to be trying both result formats, right?

Reply:
@CriesofCarrots @jstarry I think I have something workable! The issue earlier was that I didn't know we needed JS API changes to be backwards-compatible with respect to server responses since my earlier PR to solana rust client wasn't, I think I got lost in the notions of client vs server-side backwards compatibility across multiple conversations.

* I needed to add the `xyzAndContext()` methods and keep because the JS tests were too gnarly to port to the new value format *quickly*
* If someone has opinions about those methods, I don't object to anyone diving in
* I've incorporated all your other feedback (backwards compatibility etc)
* this should unblock @jackcmay and give him a drop-in version of this package to work with

Reply:
Sweet, looks close